apollo-federation/src/sources/connect/fetch_dependency_graph/mod.rs

```diff
@@ -93,10 +96,231 @@ impl FetchDependencyGraphApi for FetchDependencyGraph {
     fn add_path(
         &self,
         _query_graph: Arc<FederatedQueryGraph>,
-        _source_path: source::fetch_dependency_graph::Path,
-        _source_data: &mut source::fetch_dependency_graph::Node,
+        source_path: source::fetch_dependency_graph::Path,
+        source_data: &mut source::fetch_dependency_graph::Node,
     ) -> Result<(), FederationError> {
-        todo!()
+        // Since we are handling connect code, we should make sure that we actually have connect data
+        let source::fetch_dependency_graph::Path::Connect(source_path) = source_path else {
+            return Err(FederationError::internal("expected connect path"));
+        };
+        let source::fetch_dependency_graph::Node::Connect(source_data) = source_data else {
+            return Err(FederationError::internal("expected connect path"));
+        };
+
+        // We should be at the same merge level
+        // Note: This should be a fast comparison (pointer-level)
+        if source_path.merge_at != source_data.merge_at {
+            return Err(FederationError::internal(
+                "expected to have matching merge elements",
+            ));
+        }
+
+        // The given edge should be one that connects to both the source_path and source_data
+        // Note: Comparison of two numbers is fast
+        if source_path.source_entering_edge != source_data.source_entering_edge {
+            return Err(FederationError::internal(
+                "expected to have matching entering edges",
+            ));
+        }
+
+        // If we don't have any field selections in the source_path, then there is nothing to do.
+        let Some(source_path_field) = source_path.field else {
+            return Ok(());
+        };
+
+        // Enforce that the field contains properties shared with the source data
+        if source_path_field.response_name != source_data.field_response_name {
+            return Err(FederationError::internal(
+                "expected path and source data to have the same field name",
+            ));
+        }
+        if source_path_field.arguments != source_data.field_arguments {
+            return Err(FederationError::internal(
+                "expected path and source data to have the same field arguments",
+            ));
+        }
+
+        // Ensure that we have a selection, inserting an initial value if not.
+        let selection =
```

Comment:
is there an opportunity to move JSONSelection manipulation code into the json_selection module and have a bunch of small unit tests?

Reply:
This should be future work so that we aren't blocked for now.

Reply:
@nicholascioli I strongly disagree with not writing unit tests to avoid being blocked, because you're adding risk down the line (when deadlines are even nearer) that we won't be able to ship what we think we've built, because we didn't build the necessary confidence as we built it. Please rethink this tradeoff, because it's really not as pragmatic as you may think.

Here's an example of a case where you would have caught a bug if you'd tested the code in question: https://github.com/apollographql/router/pull/5206#discussion_r1613662011

Reply:
Thanks. Issue is fixed in https://github.com/apollographql/router/pull/5248